Add transform function prop

rmccue opened the pull request:
First pass at #31.

This allows you to do something like:
```js
const transform = (node, children) => {
    if (node.nodeName === 'UL' && node.className === 'Tasklist') {
        return <Tasklist>{children}</Tasklist>;
    }
};
const Content = props => <Interweave {...props} transform={transform} />
```

Internally, the parser is doing a depth-first parse (by calling `this.parseNode` ahead of `transform()`). This means that you receive `children` with the already-parsed React nodes; whether this is desirable I'm not certain.

(I've not really written TypeScript before, and didn't test with it now, so not certain it's fully correct. Apologies if it's not!)

rmccue commented:
> This means that you receive `children` with the already-parsed React nodes; whether this is desirable I'm not certain.

The way react-html-parser handles this is by having the `transform` function [call `convertNodeToElement` internally](https://github.com/wrakky/react-html-parser#function-convertnodetoelementnode-index-transform); we could pass through `this` and allow `transform` to call `parser.parseNode(node, config)` itself instead.

rmccue commented:
(Tests are failing on master, there are a few failing where there's now an extra `transform={null}` default prop.)

milesj commented on packages/interweave/src/Parser.js:
```diff
@@ -441,6 +441,21 @@ export default class Parser {
+        } else if (transformed !== undefined) {
+          this.keyIndex += 1;
+          content.push((
+            <React.Fragment key={this.keyIndex}>
```
What is the fragment for?

rmccue replied:
The fragment allows us to attach the `key` for the array, but otherwise does nothing. We could alternatively `React.cloneElement` on the result, but I'd rather avoid that.

milesj replied:
I'd rather not use a fragment as it would be a breaking change.

rmccue replied:
How so? This would be opt-in behaviour, since if you don't declare `transform` you'll end up with the same tree as always, so it wouldn't break any existing behaviour.

rmccue replied:
Oh right, upgrading React would be breaking 👍 Will update.

milesj commented on packages/interweave/src/Parser.js:
```diff
@@ -441,6 +441,21 @@ export default class Parser {
+        // Apply transformation if available
+        const children = this.parseNode(nextNode, config);
+        const transformed = transform ? transform(nextNode, children, config) : null;
+        if (transformed === null) {
```
Is this a way to programmatically remove nodes?

rmccue replied:
Yep, following the same pattern as react-html-parser's return format: `undefined` = use default handling, `null` = skip node, anything else = insert to tree

milesj replied:
As it stands, this would abort early without rendering the children IF `transform` is not defined.

rmccue replied:
Oh good point, that should be `transform ? ... : undefined` instead

milesj commented on packages/interweave/src/Parser.js:
```diff
@@ -441,6 +441,21 @@ export default class Parser {
+        const transformed = transform ? transform(nextNode, children, config) : null;
+        if (transformed === null) {
+          return;
+        } else if (transformed !== undefined) {
```
I prefer `typeof != undefined` checks.

rmccue commented:
Added tests and fixed the stuff broken by adding the new `transform` prop, but the other tests are all still broken due to `master` being broken.

milesj commented:
I'll take a look at master soon and then you can rebase.

rmccue commented:
Fantastic, thanks! Do you want me to squash these commits after rebasing, or would you rather do it when you merge?

milesj commented:
Doing it through the PR is fine.

milesj commented:
Master should be fixed now.

rmccue commented:
Tests are currently breaking on a bunch of things with "Compared values have no visual difference."

My guess is that now we're doing depth-first instead of leaf-first parsing (because `const children = this.parseNode()` is called before the actual element itself), this is changing the keying around. Trying to confirm this hypothesis.

rmccue commented:
I hadn't realised this was using Flow rather than TypeScript, my bad. Not quite sure how to rectify the final type error here, which is that `React$Node` is wider than what `React.cloneElement` allows. Using a fragment would allow this, but per above we can't use that. `React.isValidElement` is apparently useless for this too.

rmccue commented:
(The regular tests now passing; https://github.com/milesj/interweave/pull/32/commits/d5326ac819f51fa5c8d5c62fcc569658b6e5f3a7 is a bit of a hack, but ensures that keys are more predictable for users. In reality, this is only a concern for testing, since IMO keys are opaque and an implementation detail.)

milesj commented:
Try moving your transform logic below the `disableWhitelist` if block. That should probably happen anyways.

rmccue commented:
It was actually intentional to have it there: IMO, the explicit return of the transform should override whether the elements are whitelisted. For example, I may want to only whitelist `P` and `H2`, but if I find a specific thing to transform, I may want to return a `BLOCKQUOTE`.

milesj commented:
Right, but since you moved the keyIndex setter, the keys are probably out of order now. You may need to do something like this.

(Ghetto pseudo code)

```
if (this.props.transform) {
  parse child
  update key
  run transform
}
```

Then reset everything after like 450.

rmccue commented:
Right, makes sense. The only problem with moving this into a branch is that we'd need to parse the children twice (as they depend on the keyIndex), which would be inefficient. We'd end up with something like this (pseudo-code):

```js
if (this.props.transform) {
  this.keyIndex += 1;
  const key = this.keyIndex;
  const children = this.parseNode(nextNode, config);
  const transformed = transform(nextNode, children, config);
  if (transformed) {
    content.push(React.cloneElement(transformed, {key}));
    return;
  }

  // Reset keyIndex
  this.keyIndex = key - 1;
}

if ( disableWhitelistAndOthers ) {
  this.keyIndex += 1;
  content.push(<Element>{this.parseNode(nextNode, config)}</Element>);
}
```

To avoid calling `this.parseNode` potentially twice, we need to gather `children` early, which means we need to adjust the keyIndex before calling it for back-compat consistency.

milesj commented:
You could put a `let children = null` above the transform, then set `children` in the transform, then check for it during the `disableWhitelistAndOthers` block, like:

```
content.push(<Element>{children || this.parseNode(nextNode, config)}</Element>);
```

That way we only parse once.

rmccue commented:
Sorted, now just failing the Flow type check, which I'm not sure how to fix. We could specify that `transform` must return a `React$Element` instead of `React$Node` I guess? (Using `React.isValidElement` does not help.)

milesj commented:
I've found `React.cloneElement` to be buggy in regard to Flow. You can just throw a `$FlowIgnore` on it. I'm most likely going to convert this to Typescript at some point in the future.

rmccue commented:
Ignored that erroring line, and rebased off master. Also added some documentation to the README for `transform`; I placed this under the HTML Parsing section rather than a top-level section like Filters/Matchers since it's an advanced use case.
